README
Win32-EnvProcess version 0.05
================================

Perl extension to set or get environment variables from
other processes on Windows only

SYNOPSIS

    use Win32::EnvProcess (:all);
    
    my $result = SetEnvProcess($pid, env_var_name, [value], ...);
    
    my @values = GetEnvProcess($pid, [env_var_name, [...]]);
    
    my @pids   = GetPids([exe_name]);
    
INSTALLATION

A C/C++ compiler is required to build this module.

To install this module type the following:

   perl Makefile.PL
   nmake
   nmake test  (see TESTS below)
   nmake install

   If using the GNU toolset (for example with Strawberry Perl), 
   use dmake instead of nmake.
   
   * NON-STANDARD step *
   ---------------------
   The method used for communication with the alien process is called
   "DLL injection", which attaches a DLL to it.  Code is executed
   When the DLL is attached to the process, or a new thread is created.
   
   The DLL is named EnvProcessDll.dll does not contain any perl components;
   
   It must be copied to a directory (folder) that is on the load path
   OF THE TARGET PROCESS.  If you are not sure where that may be, try C:\Perl\bin
   (we presumably have perl installed, and it is probably in everyone's path).
   This is done by the tests by default.
   
   copy .\blib\arch\auto\Win32\EnvProcess\EnvProcessDll.dll some-directory
   
   You do not have to use the command-line for the copy, drag-and-drop with Windows
   Explorer is probably easier.
   
   * Visual Studio 2005 issues *
   -----------------------------
   The "manifest" system on this compiler needs to be taken into account by MakeMaker.
   Should you see error messages like: 
      "This application has failed to start because MSVCR80.dll was not found..."
   or
      "Runtime Error! R6034 An application has made an attempt to load the C runtime..."
   
   Then you probably need to patch MakeMaker:
   
   --- lib\ExtUtils\MM_Win32.pm.save Mon Jun 18 14:52:14 2007
   +++ lib\ExtUtils\MM_Win32.pm Mon Sep 17 11:00:19 2007
   @@ -341,6 +341,8 @@
   push(@m,
   q{ $(LD) -out:$@ $(LDDLFLAGS) }.$ldfrom.q{ $(OTHERLDFLAGS) }
   .q{$(MYEXTLIB) $(PERL_ARCHIVE) $(LDLOADLIBS) -def:$(EXPORT_LIST)});
   + push @m, '
   + if exist $@.manifest mt.exe -nologo -manifest $@.manifest -outputresource:$@;2';
   }
   push @m, '
   $(CHMOD) $(PERM_RWX) $@
   
   This trivial patch teaches makemaker to embed manifests in all cases that they are
   created as a side-effect from compiling under VC 2005. 

   See  http://www.perlmonks.com/?node_id=678699
   
   Many thanks to Yoann Pearson for his help in resolving this, and, as always, to perlmonks.

TESTS

Currently assume you are launching the tests from cmd.exe.  This will not work if
another shell is being used.  


DEPENDENCIES

None

COPYRIGHT AND LICENCE

Copyright (C) 2008 by C. B. Darke

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Win32::EnvProcess.

0.01  Fri Dec 21 14:23:00 2007
	- original version; created by h2xs 1.23 with options
		-b 5.6.1 -n Win32::EnvProcess

0.02  Fri Feb 15 10:15:00 2008
        - changed tests to allow for multiple cmd.exe sessions

0.03  Wed Mar 26 11:03:00 2008
	- removed dependency on Psapi.lib (specifically for Strawberry Perl)
	- default process id for functions is now the parent process id.
	- additional tests for new features
	
0.04  Thu June 19 11:10:00 2008
	- minor corrections to remove warnings in gcc (for Strawberry Perl)
	- improvements to free remote memory and fix a small memory leak
	- defaulting the name of the env.var. in GetEnvProcess now returns a
	  list of the first 127 environment variables and their values
	- additional tests for new features
	- README documentation of MakeMaker patch for Visual Studio 2005

0.05  Sun June 22 15:30:00 2008
	- Clarified in the README file that a C/C++ compiler is required
	- Fixed a bug (from 0.04) where an nonexistent variable returned the
	  PID from GetEnvProcess.
	- Fixed a bug where a large environment block resulted in no variables
	  being returned by GetEnvProcess.  An error message is now returned.
	- Improved error reporting.  Nonexistent environment variables now set
	  $^E to 603 ERROR_ENVVAR_NOT_FOUND.  Other errors set relevant values.
	- Improved POD concerning error numbers
	- Mutex was being released too early in SetEnvProcess and DelEnvProcess
	- Adapted tests for changes

Makefile.PL
use 5.006001;
use ExtUtils::MakeMaker;

package MY;
use Config;

sub postamble {

    # Which compiler?
    if ($Config{cc} eq 'gcc') {
    
        return <<'MAKE_GNU_FRAG';
 
Dll: EnvProcessDll.c
	$(CC) -c EnvProcessDll.c 
	$(CC) -shared -o .\$(INST_ARCHLIB)\auto\$(FULLEXT)\EnvProcessDll.$(SO) EnvProcessDll.o
MAKE_GNU_FRAG

    }
    else {
        
        return <<'MAKE_FRAG';
 
Dll: EnvProcessDll.c
	$(CC) $(CCFLAGS) EnvProcessDll.c  /link /OUT:.\$(INST_ARCHLIB)\auto\$(FULLEXT)\EnvProcessDll.$(SO) -dll
MAKE_FRAG

    }
}

sub top_targets {
  my $inherited = shift->SUPER::top_targets(@_);
  $inherited =~ s/(\npure_all\s+::.+)/$1 Dll/;
  return $inherited;
}

1;

package main;

die 'OS unsupported' if $^O ne 'MSWin32';

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Win32::EnvProcess',
    VERSION_FROM      => 'lib/Win32/EnvProcess.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Win32/EnvProcess.pm', # retrieve abstract from module
       AUTHOR         => 'C. B. Darke <clive.darke@talk21.com>') : ()),
    LIBS              => [ ], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);

if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Win32::EnvProcess',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}

1;



EnvProcessDll.c
// ----------------------------------------------------------------------------------
//  EnvProcessDll.dll
// Used by the Perl XS module Win32::EnvProcess
// Version 0.04
// ----------------------------------------------------------------------------------

#include <windows.h>
#include <stdio.h>

#include "EnvProcess.h"

void MapEnv    (void);
void SetEnv    (char *p);
void GetEnv    (char *p);
void DelEnv    (char *p);
void GetAllEnv (char *p);

/*
   It would be nicer here to avoid the temporary fixed-length arrays.
   The problem is that the target process might be linked to a
   single threaded RTL, so use of the heap would be dangerous.
*/

//--------------------------------------------------------------------
//   DLL Main
//--------------------------------------------------------------------

BOOL WINAPI DllMain(HINSTANCE hInstDLL, DWORD fdwReason, LPVOID LPReserved)
{

   switch(fdwReason)
   {
     case DLL_PROCESS_ATTACH:
     	/* This will be executed the first time we attach to a process */
        MapEnv();

        // Did not really want to do this, but can see no alternative
         FreeLibraryAndExitThread(hInstDLL, 0);

        break;
     case DLL_THREAD_ATTACH:
        /* This will be executed each time a new thread is created */
        MapEnv();
        break;
     case DLL_PROCESS_DETACH:
        break;
     case DLL_THREAD_DETACH:
        break;
     default:
        break;
  }

  return 1;
}

// -------------------------------------------------------------------

void MapEnv (void)
{
    char *p = NULL;

    HANDLE hMap = OpenFileMapping (FILE_MAP_WRITE, FALSE, FMONAME);

	if (hMap)
	   p = (char *)MapViewOfFile (hMap, FILE_MAP_WRITE, 0, 0, 0);

    if (p == NULL) {
       CloseHandle (hMap);
       return;
    }

    /* Get the command */
    if (p[0] == SETCMD)
       SetEnv(p);
    else
    if (p[0] == GETCMD)
       GetEnv(p);
    else
    if (p[0] == DELCMD)
       DelEnv(p);
    else
    if (p[0] == GETALLCMD)
       GetAllEnv(p);
    else
       p[0] = INVALID_CMD;

    UnmapViewOfFile (p);
    CloseHandle (hMap);
}

// -------------------------------------------------------------------

void SetEnv (char *p)
{
    char *p2;
    char *pszName;
    char *pszValue;
    int count;
    int i;
    BOOL bResults[MAXITEMS] = {0};

    /* Get the number of pairs */
    count = (int)p[1]/2;

    p2 = p + 2;
    for (i = 0; i < count; i++) {
       pszName = p2;
       pszValue = pszName + strlen(pszName) + 1;
       bResults[i] = SetEnvironmentVariable (pszName, pszValue);
       p2 = pszValue + strlen(pszValue) + 1;
    }

    /* Return the results */
    p2 = p + 2;
    memcpy (p2, bResults, count * sizeof(BOOL));

}  /* SetEnv */

// -------------------------------------------------------------------

void DelEnv (char *p)
{
    char *p2;
    char *pszName;
    int count;
    int i;
    BOOL bResults[MAXITEMS] = {0};

    /* Get the number of names */
    count = (int)p[1];

    p2 = p + 2;
    for (i = 0; i < count; i++) {
       pszName = p2;
       bResults[i] = SetEnvironmentVariable (pszName, NULL);
       p2 += strlen(pszName) + 1;
    }

    /* Return the results */
    p2 = p + 2;
    memcpy (p2, bResults, count * sizeof(BOOL));

}  /* DelEnv */

// -------------------------------------------------------------------
/*
   It would be nicer here to avoid the temporary array Names.
   The problem is that the target process might be linked to a
   single threaded RTL, so use of the heap would be dangerous.
*/

void GetEnv (char *p)
{
    char Names[MAXSIZE] = {0};
    char *p2;           /* pointer within FMO */
    char *pszName;      /* Pointer within local copy */
    int count;
    int i;
    DWORD dwlen;

    /* Get the number of variable names */
    count = (int)p[1];

    /* Make a copy of the names */
    p2 = p + 2;
    pszName = Names;

    for (i = 0; i < count; i++) {
	    strcpy (pszName, p2);

        dwlen = strlen(p2) + 1;
	    pszName += dwlen;
	    p2 += dwlen;
	}

    /* Get the values */
    p2 = p + 2;
    pszName = Names;

    dwlen = MAXSIZE - 2;  /* How many bytes do we have remaining? */

    for (i = 0; i < count; i++) {
       DWORD dwSize = GetEnvironmentVariable (pszName, p2, dwlen);

	   if (dwSize > dwlen) {
		   *p2 = '\0';
		   p[0] = VALUE_TOO_BIG;
		   break;
       }

	   if (dwSize == 0 ) {
		   *p2 = '\0';

		   if (GetLastError() == ERROR_ENVVAR_NOT_FOUND)
		       p[0] = ENVVAR_NOT_FOUND;
       }

	   pszName += strlen(pszName) + 1;
	   p2      += dwSize + 1;
       dwlen   += dwSize + 1;
    }


}  /* GetEnv */

// -------------------------------------------------------------------

void GetAllEnv (char *p)
{
    char *p2;          /* pointer within FMO */
    int count = 0;     /* Number of items found */
    size_t len;        /* Number of bytes remaining */

    LPTSTR lpszVariable;
    LPVOID lpvEnv;

    // Get a pointer to the environment block.

    lpvEnv = GetEnvironmentStrings();

    // If the returned pointer is NULL, exit.
    if (lpvEnv == NULL) {
        p[0] = ENVVAR_NOT_FOUND;
        p[1] = '\0';
        return;
    }

	/* Get the values */
	p2 = p + 2;

	len = MAXSIZE - 2;  /* How many bytes do we have remaining? */

    // Variable strings are separated by NULL byte, and the block is
    // terminated by a NULL byte.

	for (lpszVariable = (LPTSTR) lpvEnv; *lpszVariable; /* no-op */)
	{
	    size_t dwSize = strlen(lpszVariable);

	    if (dwSize > len) {
    	    *p2 = '\0';
    	    p[0] = VALUE_TOO_BIG;
    	    break;
        }

        if (count >= MAXITEMS / 2) {
		    p[0] = ENV_TOO_MANY;
		    break;
	    }

	    if (dwSize == 0 ) {
		    *p2 = '\0';
        }
        else {
	        strcpy (p2, lpszVariable);
	        lpszVariable += dwSize + 1;
	        count++;
	    }

	    p2  += dwSize + 1;
        len += dwSize + 1;
    }

    p[1] = count;
    FreeEnvironmentStrings(lpvEnv);

}   // GetAllEnv

// -------------------------------------------------------------------



lib/Win32/EnvProcess.pm
package Win32::EnvProcess;

use 5.006001;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. 
# This allows declaration	use Win32::EnvProcess ':all';

our %EXPORT_TAGS = ( 'all' => [ qw(SetEnvProcess GetEnvProcess DelEnvProcess GetPids
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} }, 
                   'SetEnvProcess', 'GetEnvProcess', 'DelEnvProcess', 'GetPids' );

our @EXPORT = qw(
	
);

our $VERSION = '0.05';

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&Win32::EnvProcess::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
	no strict 'refs';
	# Fixed between 5.005_53 and 5.005_61
#XXX	if ($] >= 5.00561) {
#XXX	    *$AUTOLOAD = sub () { $val };
#XXX	}
#XXX	else {
	    *$AUTOLOAD = sub { $val };
#XXX	}
    }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('Win32::EnvProcess', $VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Win32::EnvProcess - Perl extension to set or get environment variables from
other processes

=head1 SYNOPSIS

  use Win32::EnvProcess qw(:all);
  
  use Win32::EnvProcess qw(SetEnvProcess); 
  my $result = SetEnvProcess($pid, env_var_name, [value], [...]);
      
  use Win32::EnvProcess qw(GetEnvProcess);
  my @values = GetEnvProcess($pid, [env_var_name, [...]]);
  
  use Win32::EnvProcess qw(DelEnvProcess);
  my $result = DelEnvProcess($pid, env_var_name, [...]);
  
  use Win32::EnvProcess qw(GetPids);
  my @pids = GetPids([exe_name]);


=head1 DESCRIPTION

This module enables the user to alter or query an unrelated process's 
environment variables.

Windows allows a process with sufficient privilege to run code in another 
process by attaching a DLL.  This is known as "DLL injection", and is used here.  

=head1 NON-STANDARD INSTALLATION STEP

The DLL used for injection is named EnvProcessDll.dll, and does not contain 
any perl components.

It must be copied to a directory (folder) that is on the load path
OF THE TARGET PROCESS.  If you are not sure where that is, try C:\Perl\bin
(we presumably have perl installed, and it is probably in everyone's path).
This is done by the tests by default.  Failing that, WINDOWS\system32 should work.

copy .\blib\arch\auto\Win32\EnvProcess\EnvProcessDll.dll some-directory

You do not have to use the command-line for the copy, drag-and-drop with 
Windows Explorer is probably easier.

=head1 EXPORT

None by default.

=head2 SetEnvProcess

my $result = SetEnvProcess($pid, env_var_name, [value], ...);

$pid:  		The process identifier (PID) of the target process.
		If set to 0 (zero) the parent process identifier is used
env_var_name:	The name of the environment variable to be set
value:		The value of the preceeding environment variable

Set one or more environment variables in another process.

The env_var_name/value pairs may be repeated as a list or, more
conviently as a hash, with the keys being the environment variable 
names.  You should avoid changing critcal process specific variables
such as USERNAME.  Also be aware that there is no guarantee that the
program will actually use the new value.  For example it may have
already read the value on start-up and may be holding it internally,
the perl interpreter is such a program (see C<Interaction with perl scripts> below).

If an odd number of items is supplied in the list, the final variable 
specified will have no value.  Note that this is not the same as deleting
a variable.

Returns: the number of environment variables changed.  Error information
will be available in $^E (See below).

=head2 GetEnvProcess

my @values = GetEnvProcess($pid, [env_var_name, [...]]);

$pid:  		The process identifier (PID) of the target process.
		If set to 0 (zero) the parent process identifier is used
env_var_name:	The name[s] of the environment variable[s] to be read.
		If not supplied then as many environment variables as possible
		are returned (see C<LIMITATIONS> below).

Get one or more environment variable values from another process.

Returns undef on error, error information will be available in $^E.
If one or more environment variable name are specified, a list of environment 
variable values is returned.  Note that these may be empty strings if no value is set.
If a requested variable does not exist then an empty string is returned in it's place
in the list, and $^E will be set to 203 (ERROR_ENVVAR_NOT_FOUND).

If no environment variable names are specified then the environment block is returned (subject to LIMITATIONS) in the form of a list.  
Items are of the form variable=value and may be extracted into a hash as follows:
  
  @values = GetEnvProcess (0);
  
  my %proc_env;
  for (@values) {
      my ($key, $value) = split /=/;
      $proc_env{$key} = $value;
  }

=head2 DelEnvProcess

my $result = DelEnvProcess($pid, env_var_name, ...);

$pid:  		The process identifier (PID) of the target process.
		If set to 0 (zero) the parent process identifier is used
env_var_name:	The name of the environment variable to be deleted

Delete one or more environment variables in another process.

The env_var_name may be repeated as a list.  You should avoid deleting 
critcal process specific variables such as USERNAME and PATH.  
Also be aware that the program may have already read the value on 
start-up and may be holding it internally, the perl interpreter is such 
a program (see C<Interaction with perl scripts> below).

Returns: the number of environment variables deleted.  Error information
will be available in $^E (See below).

=head2 GetPids

my @pids = GetPids([exe_name]);

exe_name:	The basename of an executable file (including the '.exe')

If supplied, a case-insensitive match is done between the exe name and running 
processes.  If not supplied, then the identifier of the parent process is returned.

Returns: A list of process-ids of processes running the supplied exe, or zero on 
error.  This function is provided to assist in obtaining the required process id.  

Security may prevent some processes from being queried.  When this is the case it 
is unlikely that we have the permissions to manipulate the environment block.  
The .exe files registered with a process may be in 8.3 format, and currently 
this function makes no attempt to resolve a 'long name' to short.

=head1 $^E ($EXTENDED_OS_ERROR) VALUES

The most likely causes of certain Win32 errors are given below:

=head3 8	Not enough storage is available to process this command.

The total data for the environment variables exceeds C<LIMITATIONS>.
In this case as much processing as possible will be done.

=head3 14	Not enough storage is available to complete this operation.

The number of environment variables exceeds C<LIMITATIONS>

=head3 87	The parameter is incorrect.	

The supplied PID does not exist

=head3 183	Cannot create a file when that file already exists.

Under most circumstances this error may be ignored when correct results are returned.  
It generally means that more than one process (or thread) is using this module.

=head3 186 	The flag passed is not correct.

Internal corruption of the FMO, please contact the author

=head3 203	The system could not find the environment option that was entered

The requested environment variable does not exist


=head1 LIMITATIONS

Total size of variable names and values: 8192 (increased at v.0.05).

Total number of environment variables  :  254 (increased at v.0.05).

These are artificial limits and may be made more flexible in a future release.   

Locking: the entire sequence is serial because a named FMO is used.  Creating 
the File Mapping Object(FMO), writing to it, and running the DLL in the other 
process, is protected by a Mutex.  It is therefore possible that calls may block.

=head2 Interaction with perl scripts

If the target process is a perl script, then note that the script will not 'see'
the new variable or value through C<%ENV>.  This hash is set at process creation time,
and not directly updated through these functions.  A perl script can however inspect
its own environment using C<GetProcessEnv> with the first argument set to C<$$>.


=head1 SEE ALSO

Win2::Env, 
Env::C

=head1 AUTHOR

C. B. Darke, clive.darke@ talk21.com
open to suggestions for improvements.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by C. B. Darke

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6.1 or,
at your option, any later version of Perl 5 you may have available.


=cut
